### PR TITLE
Rebalances APC construction (Fixes #7298)

### LINKED ISSUE #7298: APC building rebalance B:20
### Environment
- Game Version: Staging 21081715
- OS Version: Arch Linux

### Detailed description of bug

right now building a new APC just requires 2 metal and placing in on a wall, while on the TG it needs as shown here:
APC

for how easy it is to make a new APC and with the doors V2 now on all stations it can lead to cheesing the apc to open any door on station.

https://user-images.githubusercontent.com/80334192/130170768-fbe955cd-28b7-49a8-a920-bb73c9e753b4.mp4

### PR DESCRIPTION
**Pull request by AtnerNT**
### Purpose
Adds new APC construction and deconstruction steps to APC's to prevent related door exploits. Aims to fix #7298 

### Steps to Construct:
The new steps to APC construction are as follows:

1) Place an APC frame on a wall

Step1

2) Add cables to the APC frame (x5)

Step2

3) Add a power control module to the frame

Step3

4) Add a power cell (Any level of power cell works- but doing so has no effect on the performance of the APC)

Step4

5) (Wrench) Secure the APC's cover in place

Step5

6) (Screw Driver) Screw in the electronics and complete construction

Step6

### Steps to Deconstruct:
The new steps to APC deconstruction are as follows:

1) (Screw driver) Unscrew the electronics, doing so will cause electrocution if not wearing proper insulative equipment

DisStep1

2) (Crowbar) Pry off the cover

DisStep2

3) (Crowbar) Remove the power control module and power cell

DisStep3

4) (Wire cutters) Remove the cables from the frame

DisStep4

5) (Welding Tool) Dismantle the frame

DisStep5

### Changelog:

CL: [Balance] Adds new construction and deconstruction steps for APC's aswell as added electrocution if attempting to disconnect the APC without gloves

### CODE REVIEW

File: UnityProject/Assets/APCFrame.cs [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+			powerCellSlot = GetComponent<ItemStorage>().GetIndexedItemSlot(1);
+			stateful = GetComponent<Stateful>();
+			objectBehaviour = GetComponent<ObjectBehaviour>();
+
+			if (!CustomNetworkManager.IsServer) return;
+
+			integrity = GetComponent<Integrity>();
+			spriteHandler = GetComponentInChildren<SpriteHandler>();
+
+			integrity.OnWillDestroyServer.AddListener(WhenDestroyed);
```
Comment (ThatDan123) [inline UnityProject/Assets/APCFrame.cs]: Could you move that to OnEnable and add a remove in OnDisable

File: UnityProject/Assets/APCFrame.cs [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+			if (CurrentState != powerCellAddedState)
+			{
+				stateful.ServerChangeState(initialState);
+			}
+		}
+
+		public override void OnStartClient()
+		{
+			base.OnStartClient();
+		}
```
Comment (ThatDan123) [inline UnityProject/Assets/APCFrame.cs]: unused

File: UnityProject/Assets/Scripts/Objects/Engineering/APC.cs [hunk 3]
```diff
@@ -80,8 +83,14 @@ private void SyncVoltage(float oldVoltage, float newVoltage)
...
 		private void Awake()
 		{
+			powerControlSlot = GetComponent<ItemStorage>().GetIndexedItemSlot(0);
+			powerCellSlot = GetComponent<ItemStorage>().GetIndexedItemSlot(1);
+
 			electricalNodeControl = GetComponent<ElectricalNodeControl>();
 			resistanceSourceModule = GetComponent<ResistanceSourceModule>();
+			integrity = GetComponent<Integrity>();
+
+			integrity.OnWillDestroyServer.AddListener(WhenDestroyed);
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/Objects/Engineering/APC.cs:95]: same here

**Comment by ThatDan123** [comment]
You also need to open the networkmanager prefab and press then button to fill the prefab list and then commit the networkmanager change please. That'll fix the tests.

Also Codacy has picked something up

**Comment by AtnerNT** [comment]
God dammit- well it's getting late over here- I'll figure out why the tests are failing and try fix them tomorrow

File: UnityProject/Assets/Scripts/Core/Input System/WallMountHandApplySpawn.cs [hunk 4]
```diff
@@ -2,13 +2,11 @@
 using System.Collections;
 using System.Collections.Generic;
 using UnityEngine;
-using Systems.Electricity;
 
 // TODO: namespace me
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Core/Input System/WallMountHandApplySpawn.cs:6]: maybe it is a good time to do it

File: UnityProject/Assets/Scripts/Objects/Engineering/APC.cs [hunk 5]
```diff
@@ -487,5 +526,62 @@ private void AddDeviceFromScene(APCPoweredDevice poweredDevice)
...
+		public void WhenDestroyed(DestructionInfo info)
+		{
+			// rare cases were gameObject is destroyed for some reason and then the method is called
+			if (gameObject == null) return;
+
+			Inventory.ServerSpawnPrefab(powerControlModule, powerControlSlot, ReplacementStrategy.Cancel);
+			Inventory.ServerSpawnPrefab(powerCell, powerCellSlot, ReplacementStrategy.Cancel);
+
+			SpawnResult frameSpawn = Spawn.ServerPrefab(APCFrameObj, SpawnDestination.At(gameObject));
+			if (!frameSpawn.Successful)
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Objects/Engineering/APC.cs]:
```suggestion
			if (frameSpawn.Successful == false)
```

**Comment by corp-0** [comment]
```
MachineFrame (PrefabTracker) or APCFrame (PrefabTracker) NEEDS to be committed with it's new Forever IDat Tests.Asset.NetworkManagerTests.SpawnableListTest () [0x001dc] in /github/workspace/UnityProject/Assets/Tests/Asset/NetworkManagerTests.cs:81
```
Ask Bod how to generate the prefab ID

File: UnityProject/Assets/Scripts/Objects/Engineering/APC.cs [hunk 6]
```diff
@@ -487,5 +526,62 @@ private void AddDeviceFromScene(APCPoweredDevice poweredDevice)
 		}
 
 		#endregion
+
+		public bool WillInteract(HandApply interaction, NetworkSide side)
+		{
+			return Validations.HasUsedItemTrait(interaction, CommonTraits.Instance.Screwdriver);
+		}
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/Objects/Engineering/APC.cs]:
```suggestion
		public bool WillInteract(HandApply interaction, NetworkSide side)
		{
			if (DefaultWillInteract.Default(interaction, side) == false) return false;
			
			return Validations.HasUsedItemTrait(interaction, CommonTraits.Instance.Screwdriver);
		}
```

**Review by ThatDan123 — APPROVED** [review]

**Review by corp-0 — APPROVED** [review]

**Comment by PerfectTangent** [comment]
@AtnerNT hmu on discord @jack Trachea if you would like to collect on the bounty

**Pull request merged**